packages/block-library/src/template-part/edit/index.js

```diff
@@ -27,44 +27,68 @@ export default function TemplatePartEdit( {
 	setAttributes,
 	clientId,
 } ) {
-	const initialPostId = useRef( _postId );
 	const initialSlug = useRef( slug );
 	const initialTheme = useRef( theme );
+	const initialContent = useRef();
 
 	// Resolve the post ID if not set, and load its post.
 	const postId = useTemplatePartPost( _postId, slug, theme );
 
-	// Set the post ID, once found, so that edits persist,
-	// but wait until the third inner blocks change,
-	// because the first 2 are just the template part
-	// content loading.
-	const { innerBlocks } = useSelect(
+	// Set the postId block attribute if it did not exist,
+	// but wait until the inner blocks have loaded to allow
+	// new edits to trigger this.
+	const { innerBlocks, expectedContent } = useSelect(
 		( select ) => {
 			const { getBlocks } = select( 'core/block-editor' );
+			const entityRecord = select( 'core' ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				postId
+			);
+
 			return {
 				innerBlocks: getBlocks( clientId ),
+				expectedContent: entityRecord?.content.raw,
 			};
 		},
-		[ clientId ]
+		[ clientId, postId ]
 	);
 	const { editEntityRecord } = useDispatch( 'core' );
-	const blockChanges = useRef( 0 );
 	useEffect( () => {
-		if ( blockChanges.current < 4 ) blockChanges.current++;
-
+		// If postId (entity) has not resolved or _postId (block attr) is set,
+		// then we have no need for this effect.
 		if (
-			blockChanges.current === 3 &&
-			( initialPostId.current === undefined ||
-				initialPostId.current === null ) &&
-			postId !== undefined &&
-			postId !== null
+			postId === null ||
+			postId === undefined ||
+			( _postId !== null && _postId !== undefined )
```

Comment:
Could we simplify these with simple truthy/falsy checks, or is there a reason why we actually need to know if `null/undefined`?
I wouldn't expect `postId` to ever be either `0` or `false`... 🤔 

```js
if ( ! postId || !! _postId ) { }
```

Reply:
Thats a good point.  It was initially written with the `null/undefined` checks for whatever reason, my guess being the `postId` might be `0` thing (I can't think of any other reason?).  I have seen them in the low single digits, but im not 100% certain if one can ever be `0`. 🤔

Reply:
I'm pretty sure a post ID can't be 0, if anything because the DB is not set that way (e.g. inserting `0` as primary key is automatically transformed into the next available key value, unless there are special configurations in place).

Reply:
As far as i can tell nothing should have the id `0`. 👍